src/cryptonote_core/service_node_list.cpp

```diff
@@ -312,7 +320,114 @@ namespace service_nodes
 
 
     m_rollback_events.push_back(std::unique_ptr<rollback_event>(new rollback_change(block_height, key, iter->second)));
+    uint64_t swarm_id = iter->second.swarm_id;
     m_service_nodes_infos.erase(iter);
+    check_and_process_swarm_decomission(swarm_id);
+  }
+
+  void service_node_list::check_and_process_swarm_decomission(uint64_t swarm_id)
+  {
+    size_t count = 0;
+    for (const auto& entry : m_service_nodes_infos)
+      if (entry.second.swarm_id == swarm_id)
+        count++;
+    if (count == MIN_SWARM_SIZE - 1)
+    {
+      // XXX XXX XXX XXX
+      // All nodes should fetch data from this swarm at this point only.
+      // The registered nodes will eventually disappear and after this point,
+      // it is already considered gone. It only exists to retrieve data from.
+      //
+      // XXX XXX XXX XXX
+      //         internship optimization hardfork idea:
+      //         when nodes have been decomissioned for more than 10 blocks,
+      //         move the nodes into new swarms immediately.
+    }
+  }
+
+  uint64_t service_node_list::get_new_node_swarm_id(uint64_t block_height, uint32_t index) const
+  {
+    // XXX XXX XXX improvement proposal: have the swarm id be "remembered" from
+    // the last time, if the new node was already registered before. This would
+    // have huge performance gains preventing nodes from having to resync every
+    // time they rejoin the network.
+
+    crypto::hash hash = m_blockchain.get_block_id_by_height(block_height);
+    uint64_t seed = 0;
+    std::memcpy(&seed, hash.data, sizeof(seed));
+    seed += index;
+    std::mt19937_64 mersenne_twister(seed);
+
+    std::vector<uint64_t> swarm_ids = get_swarm_ids();
+
+    if (swarm_ids.empty())
+      return 0;
+
+    const size_t swarm_index = (size_t)uniform_distribution_portable(mersenne_twister, swarm_ids.size());
+
+    return swarm_ids[swarm_index];
+  }
+
+  void service_node_list::check_and_process_swarm_overflow(uint64_t swarm_id, uint64_t block_height)
+  {
+    uint64_t count = 0;
+    for (const auto& entry : m_service_nodes_infos)
+      if (entry.second.swarm_id == swarm_id)
+        count++;
+
+    if (count > MAX_SWARM_SIZE)
+    {
+      crypto::hash hash = m_blockchain.get_block_id_by_height(block_height);
+      uint64_t seed = 0;
+      std::memcpy(&seed, hash.data, sizeof(seed));
+      seed += swarm_id;
+      std::mt19937_64 mersenne_twister(seed);
+
+      bool move_to_new_swarm = (bool)uniform_distribution_portable(mersenne_twister, 2);
+      uint64_t new_swarm = uniform_distribution_portable(mersenne_twister, UINT64_MAX);
```

Comment:
Any reason it has to be (0->U64 MAX)? Could collide with pre-existing swarms. How about making the new swarm id = (num_swarms + 1)

Reply:
because the messenger client pubkey -> swarm id mapping requires a random distribution to split the space up evenly. load balancing.

click around randomly here for what separating the space with random points look like: http://alexbeutel.com/webgl/voronoi.html

here we're doing it in 256 dimensions instead of 2, but it's the same idea (I think...)

good point about colliding with an existing swarm, you should probably put a check in there if I don't have time to do it

Reply:
(that said it's not a huge problem if one of the swarms gets an extra guy for a little while)